File: csharp-solution/E-CommerceSystem/Models/CheckoutService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace E_CommerceSystem.Models
{
  public class CheckoutService
  {
    private const double ShappingFee = 60.0;
    public static void Checkout(Customer customer,Cart cart)
    {
      if (cart.isEmpty)
        throw new InvalidOperationException("Cart is empty");
      double Subtotal = 0;
      List<IShippable> shipItems=new List<IShippable>();
      foreach (var item in cart.GetItems())
      {
        var product = item.Key;
        var quntity=item.Value;

        if (product.IsExpired())
          throw new InvalidOperationException($"{product.Name} is expired!");

        if (!product.IsAvailable(quntity))
          throw new InvalidOperationException($"{product.Name} out of stock!");

        Subtotal += product.Price * quntity;

        if(product is IShippable shippableProduct)
        {
          for (int i = 0;i<quntity;i++)
            shipItems.Add(shippableProduct);
        }

      }
      double Shipping = (shipItems.Count == 0 ? 0 : ShappingFee);
        double Amount = Subtotal + Shipping;
        if (customer.Balance < Amount)
          throw new InvalidOperationException("Insufficient balance");

        if (shipItems.Count > 0)
          ShippingService.Ship(shipItems);

        foreach (var item in cart.GetItems())
        {
          item.Key.ReduceQuantity(item.Value);
        }
        customer.Deduct(Amount);

      Console.WriteLine("\n** Checkout receipt **");
      Console.WriteLine($"{"Qty",-4} {"Product",-20} {"Price"}");

      foreach (var item in cart.GetItems())
      {
        Console.WriteLine($"{item.Value,-4} {item.Key.Name,-20} {item.Key.Price}"); 
      }

      Console.WriteLine("-------------------------------------------------");
      Console.WriteLine($"Subtotal = {Subtotal}");
      Console.WriteLine($"Shipping = {Shipping}");
      Console.WriteLine($"Amount = {Amount}");

    }
  }
}

File: java-solution/src/service/CheckoutService.java
package service;

import Model.Customer;
import Model.Product;
import interfaces.IExpirable;
import interfaces.IShippable;
import Model.cart;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class CheckoutService {
    private static final double SHIPPING_FEE = 60.0;

    public static void checkout(Customer customer, cart cart) {
        if (cart.isEmpty())
            throw new IllegalArgumentException("Cart is empty");

        double subtotal = 0.0;
        List<IShippable> shipItems = new ArrayList<>();

        for (Map.Entry<Product, Integer> item : cart.getItems().entrySet()) {
            Product product = item.getKey();
            int quantity = item.getValue();

            if (!product.isAvailable(quantity))
                throw new IllegalStateException(product.getName() + " out of stock");

           if (product instanceof IExpirable) {
               IExpirable expirable=(IExpirable) product;
               if (expirable.isExpired())
                   throw new IllegalStateException(product.getName() + " is expired");
           }

            subtotal += product.getPrice() * quantity;
            if (product instanceof IShippable) {
                IShippable s = (IShippable) product;
                for (int i = 0; i < quantity; i++) {
                    shipItems.add(s);
                }
            }
        }

        double shipping = shipItems.isEmpty() ? 0.0 : SHIPPING_FEE;
        double amount = subtotal + shipping;

        if (customer.getBalance() < amount)
            throw new IllegalStateException("Insufficient balance");

        if (!shipItems.isEmpty())
            ShippingService.ship(shipItems);

        customer.deduct(amount);

        for (Map.Entry<Product, Integer> entry : cart.getItems().entrySet()) {
            entry.getKey().reduceQuantity(entry.getValue());
        }

        System.out.println("** Checkout receipt **");
        System.out.printf("%-4s %-20s %10s%n", "Qty", "Product", "Price");
        System.out.println("------------------------------------------------------");

        for (Map.Entry<Product, Integer> entry : cart.getItems().entrySet()) {
            int qty = entry.getValue();
            Product p = entry.getKey();
            System.out.printf("%-4d %-20s %10.2f%n", qty, p.getName(), p.getPrice() * qty);
        }

        System.out.println("----------------------------------------------------------------");
        System.out.printf("%-24s %10.2f%n", "Subtotal", subtotal);
        System.out.printf("%-24s %10.2f%n", "Shipping", shipping);
        System.out.printf("%-24s %10.2f%n", "Amount", amount);

    }
}
